File: exp-13/q1-prims/solution.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int minKey(vector<int>& key, vector<bool>& mstSet, int n)
{
    int min = INT_MAX, minIndex = -1;
    for (int v = 0; v < n; v++)
        if (!mstSet[v] && key[v] < min)
        {
            min = key[v];
            minIndex = v;
        }
    return minIndex;
}

int main()
{
    int n, m;
    cin >> n >> m;

    vector<vector<int> > graph(n, vector<int>(n, 0));

    for (int i = 0; i < m; i++)
    {
        int u, v, w;
        cin >> u >> v >> w;
        graph[u - 1][v - 1] = w;
        graph[v - 1][u - 1] = w;
    }

    vector<int> parent(n, -1);
    vector<int> key(n, INT_MAX);
    vector<bool> mstSet(n, false);

    key[0] = 0;

    int totalWeight = 0;
    for (int count = 0; count < n; count++)
    {
        int u = minKey(key, mstSet, n);
        mstSet[u] = true;
        totalWeight += key[u];

        for (int v = 0; v < n; v++)
            if (graph[u][v] && !mstSet[v] && graph[u][v] < key[v])
            {
                parent[v] = u;
                key[v] = graph[u][v];
            }
    }

    for (int v = 1; v < n; v++)
        cout << "Edge " << parent[v] + 1 << "-" << v + 1
             << " weight " << key[v] << "\n";
    cout << "Total weight of MST: " << totalWeight << "\n";

    return 0;
}

File: exp-13/q1-prims/Solution.java
import java.util.Scanner;

public class Solution {

    static int minKey(int[] key, boolean[] mstSet, int n) {
        int min = Integer.MAX_VALUE, minIndex = -1;
        for (int v = 0; v < n; v++)
            if (!mstSet[v] && key[v] < min) {
                min = key[v];
                minIndex = v;
            }
        return minIndex;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();

        int[][] graph = new int[n][n];

        for (int i = 0; i < m; i++) {
            int u = sc.nextInt();
            int v = sc.nextInt();
            int w = sc.nextInt();
            graph[u - 1][v - 1] = w;
            graph[v - 1][u - 1] = w;
        }

        int[] parent = new int[n];
        int[] key = new int[n];
        boolean[] mstSet = new boolean[n];

        for (int i = 0; i < n; i++) {
            key[i] = Integer.MAX_VALUE;
            parent[i] = -1;
        }

        key[0] = 0;

        int totalWeight = 0;
        for (int count = 0; count < n; count++) {
            int u = minKey(key, mstSet, n);
            mstSet[u] = true;
            totalWeight += key[u];

            for (int v = 0; v < n; v++)
                if (graph[u][v] != 0 && !mstSet[v] && graph[u][v] < key[v]) {
                    parent[v] = u;
                    key[v] = graph[u][v];
                }
        }

        for (int v = 1; v < n; v++)
            System.out.println("Edge " + (parent[v] + 1) + "-" + (v + 1)
                    + " weight " + key[v]);
        System.out.println("Total weight of MST: " + totalWeight);
    }
}
